src/agent/src/watcher.rs

```diff
@@ -0,0 +1,487 @@
+// Copyright (c) 2021 Apple Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use std::time::SystemTime;
+
+use tokio::fs;
+use tokio::sync::Mutex;
+use tokio::task;
+use tokio::time::{self, Duration};
+
+use anyhow::{ensure, Context, Result};
+use async_recursion::async_recursion;
+use nix::mount::{umount, MsFlags};
+use slog::{error, Logger};
+
+use crate::mount::BareMount;
+use crate::protocols::agent::Storage;
+
+/// The maximum number of file system entries agent will watch for each mount.
+const MAX_ENTRIES_PER_STORAGE: usize = 8;
+
+/// How often to check for modified files.
+const WATCH_INTERVAL_SECS: u64 = 2;
+
+/// Destination path for tmpfs
+const WATCH_MOUNT_POINT_PATH: &str = "/run/kata-containers/shared/containers/watchable/";
+
+/// Represents a single file entry to be watched.
+#[derive(Debug, Clone)]
+struct File {
+    file: PathBuf,
+    modified: SystemTime,
+}
+
+/// Represents a single storage entry (may have multiple files to watch).
+#[derive(Default, Debug, Clone)]
+struct Entry {
```

Comment:
It's better to iml the Drop trait for Entry to do the clean up of the files in mount_point dir, otherwise, the trash files would be left after a container stopped.